### PR TITLE
Adds more deeplink apis

### PR DESCRIPTION
**Pull request by chunhtai**
Adds 3 more api

androidAppLinkSettings
iosBuildOptions
iosUniversalLinkSettings

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or there is a reason for not adding tests.

![build.yaml badge]

If you need help, consider asking for help on [Discord].

[Contributor Guide]: https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[build.yaml badge]: https://github.com/flutter/devtools/actions/workflows/build.yaml/badge.svg

### CODE REVIEW

File: packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/config_specific/server/_server_web.dart:440]: this should be DeeplinkApi.androidAppLinkSettings instead of apiResetDevTools. Similar for updating the logWarning calls below

File: packages/devtools_shared/CHANGELOG.md
Comment (kenzieschmoll) [inline packages/devtools_shared/CHANGELOG.md:1]: 4.1.0? This isn't a breaking change so a minor bump should be okay. Can you update the pubspec.yaml file too?

File: packages/devtools_shared/lib/src/deeplink/app_link_settings.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/app_link_settings.dart:31]: dart doc

File: packages/devtools_app/lib/src/shared/config_specific/server/_server_stub.dart
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/config_specific/server/_server_stub.dart:112]: should this be "variants" instead of "options" to match the android endpoint?

Comment (chunhtai) [reply packages/devtools_app/lib/src/shared/config_specific/server/_server_stub.dart:112]: In iOS, it is call build options. `variant` is a keyword in Android gradle build system.

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:106]: Most of this is duplicated above and in the methods below. Can we use a helper method to de-dupe?

File: packages/devtools_shared/lib/src/server/server_api.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/server/server_api.dart:479]: this block is duplicated a few times. Can we use a helper `resultOutputOrError(result)`?

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (chunhtai) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:73]: @DanTup I remembered you mentioned we should have a way to locate flutter in dart app. The issue https://github.com/Dart-Code/Dart-Code/issues/4722 is closed. Is there now a way to get the path flutter binary?

Comment (DanTup) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:73]: Yep, see https://github.com/Dart-Code/Dart-Code/issues/4722#issuecomment-1722276722 - that issue has been implemented and is available in the current pre-release version of the extension. If you're using that. the `FLUTTER_ROOT` env variable should be available.

Let me know if this doesn't seem to be the case.

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (DanTup) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:50]: `FLUTTER_ROOT` will be the SDK folder (without `bin`), so you probably need to combine these.

I'm also not sure that `workingDirectory` is always searched for the executable, so I think it should be prepended to the `executable`. So you might need something like:

```dart
    return runProcess(
      path.join(flutterRoot, 'bin', Platform.isWindows ? 'flutter.bat' : 'flutter'),
      arguments: arguments,
    );
```

I would also suggest not using `!` to assume this will always be set, in case DevTools is running in a context other than VS Code, or a version of the Dart plugin that doesn't include https://github.com/Dart-Code/Dart-Code/issues/4722 (such as the current stable version).

You should be able to test this using the pre-release extension, but let me know if that doesn't seem to be the case.

Dart pre-release versions

File: packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md
Comment (kenzieschmoll) [inline packages/devtools_app/release_notes/NEXT_RELEASE_NOTES.md:25]: The release notes are intended for user-facing changes, so let's hold off on adding this note here until we ship the deep link feature in general. The entry you added to `devtools_shared/CHANGELOG.md` is sufficient. To get the release notes check on the bots, add the label "release notes not required"

File: packages/devtools_shared/lib/src/deeplink/app_link_settings.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/app_link_settings.dart:15]: please put 'applicationId' and 'deeplinks' in static consts on this class.

File: packages/devtools_shared/lib/src/deeplink/app_link_settings.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/app_link_settings.dart:34]: nit: r/are/is

File: packages/devtools_shared/lib/src/deeplink/app_link_settings.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/app_link_settings.dart:43]: please put these strings in constants on this class

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:29]: can you add dart doc to this?

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:131]: this method can use `_runFlutterCommandAndReadOutputFile` if we add two parameters `regex` and `void Function(ProcessResult) noMatchError`

```dart
return _runFlutterCommandAndReadOutputFile(
  ['analyze', '--ios', '--list-build-options', rootPath],
  regex: _iosBuildOptionsJsonRegex,
  (result) => 'Can not parse output ${result.stdout}',
);
```

File: packages/devtools_shared/lib/src/deeplink/universal_link_settings.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/universal_link_settings.dart:20]: please put strings in static consts. Here and elsewhere.

File: packages/devtools_shared/test/server/server_api_test.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/test/server/server_api_test.dart:43]: nit: in test names, consider using the the enum values directly so the test names don't go stale:
`'${DeeplinkApi.androidBuildVariants} ...'`

here and elsewhere below

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (chunhtai) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: I added the logic to get the proper path and also handler when env var is not set

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:73]: Is it really the DevTools version that determines this? We should be more clear about which version is preventing the `FLUTTER_ROOT` environment variable from being set. In other words, is this due to the Dart SDK version, the Flutter SDK version, etc.

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:77]: can we throw a specific type of exception instead of throwing a string? e.g. maybe ProcessException here and FormatException below. Here and elsewhere where we are throwing a String

File: packages/devtools_shared/test/server/server_api_test.dart
Comment (kenzieschmoll) [inline packages/devtools_shared/test/server/server_api_test.dart:15]: nit: I think now the "deeplink api" text is redundant since `'${DeeplinkApi.androidBuildVariants}'` should output "DeeplinkApi.androidBuildVariants".

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (chunhtai) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:73]: Hi @DanTup Do you know which dart version the change will be in?

Comment (chunhtai) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart:73]: ah I see, there isn't a version we can target in this case, a lot of different cases can result in this situation. For example running the dev tool directly through a standalone dart. I think I will just mention that it can't find a flutter sdk and maybe mention it needs to set FLUTTER_ROOT environment variable

File: packages/devtools_shared/test/server/server_api_test.dart
Comment (chunhtai) [reply packages/devtools_shared/test/server/server_api_test.dart:15]: The '${DeeplinkApi.androidBuildVariants}' should be evaluate to something like  `'api/androidBuildVariants'`? so I think the deeplink api is still relevant

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (kenzieschmoll) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: Can we just use 'flutter' if we don't find FLUTTER_ROOT? Not as accurate as using FLUTTER_ROOT but we would at least not fail as easily since most users should have `flutter` on their path

**Review by kenzieschmoll — APPROVED** [review]
LGTM, left one comment though about defaulting to use 'flutter' if we can't find FLUTTER_ROOT

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (DanTup) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: > Hi @DanTup Do you know which dart version the change will be in?

Version 3.74.0 of the Dart extension for VS Code will include the change to set `FLUTTER_ROOT`. As you noted, this is specific to the server being started by VS Code though, so may not apply if it's run from the command line, etc. so if it's missing, it's worth trying to handle gracefully.

Perhaps before falling back to using `flutter` from `PATH` (which may also not exist, or be a different version if the user is using something like `fvm`), it's worth checking `Platform.resolvedExecutable` (which is the path to the Dart VM that's running) and walk up to see if it's nested inside `bin/cache/dart-sdk/bin` of a Flutter SDK and using that?

Comment (chunhtai) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: I can try to use the flutter from path.

>  it's worth checking Platform.resolvedExecutable (which is the path to the Dart VM that's running) and walk up to see if it's nested inside bin/cache/dart-sdk/bin of a Flutter SDK and using that?

If it is using the dart from the flutter sdk, it should have set the FLUTTER_ROOT I think?

Comment (DanTup) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: > If it is using the dart from the flutter sdk, it should have set the FLUTTER_ROOT I think?

I _think_ if it's using the shell script at `bin/dart` it would, but if it was launched directly using `bin/cache/dart-sdk/bin/dart` I'm not sure it would. I don't know how common each way is though (personally I don't use the shell script because `dart` on my `PATH` is always a standalone Dart SDK, but my setup is likely not the same as most users :-))

**Review by kenzieschmoll — APPROVED** [review]

File: packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart
Comment (chunhtai) [reply packages/devtools_shared/lib/src/deeplink/deeplink_manager.dart]: done

**Review by DanTup — APPROVED** [review]
The use of `FLUTTER_ROOT` looks reasonable to me. I'll defer to Kenzie for the rest, as I'm less familiar with this functionality.

**Pull request merged**
